src/pymor/algorithms/bfgs.py

```diff
@@ -0,0 +1,320 @@
+    data
+        Dict containing the following fields:
+
+            :mus:                       `list` of parameters after each iteration.
+            :foc_norms:                 |NumPy array| of the first order criticity norms
+                                        after each iteration.
+            :update_norms:              |NumPy array| of the norms of the update vectors
+                                        after each iteration.
+            :iterations:                Number of total BFGS iterations.
+            :line_search_iterations:    |NumPy array| of the number of line search
+                                        iterations per BFGS iteration.
+
+    Raises
+    ------
+    BFGSError
+        Raised if the BFGS algorithm failed to converge.
+    """
+    logger = getLogger('pymor.algorithms.bfgs')
+
+    data = {}
+
+    assert model.output_functional is not None
+    output = lambda m: model.output(m)[0, 0]
+
+    if parameter_space is None:
+        parameter_space = model.parameters.space(-np.inf, np.inf)
+
+    if initial_guess is None:
+        initial_guess = parameter_space.sample_randomly(1)[0]
+        mu = initial_guess.to_numpy()
+    else:
+        mu = initial_guess.to_numpy() if isinstance(initial_guess, Mu) else initial_guess
+
+    if error_aware:
+        assert error_criterion is not None and beta is not None and radius is not None
+        assert callable(error_criterion)
+
+    gradient = model.parameters.parse(model.output_d_mu(mu)).to_numpy()
+    current_output = output(mu)
+    eps = np.linalg.norm(gradient)
+
+    # compute norms
+    mu_norm = np.linalg.norm(mu)
+    update_norms = []
+    foc_norms = []
+    line_search_iterations = []
+    data['mus'] = [mu.copy()]
+
+    hessian = np.eye(mu.size)
+
+    first_order_criticity = output_diff = mu_diff = update_norm = 1e6
+    iteration = 0
+    while True:
+        if iteration >= miniter:
+            if output_diff < rtol:
+                logger.info(f'Relative tolerance of {rtol} for output difference reached. Converged.')
+                break
+            if mu_diff < rtol:
+                logger.info(f'Relative tolerance of {rtol} for parameter difference reached. Converged.')
+                break
+            if first_order_criticity < tol_sub:
+                logger.info(f'Absolute tolerance of {tol_sub} for first order criticity reached. Converged.')
+                break
+            if error_aware:
+                if error_criterion(mu, current_output):
+                    logger.info(f'Output error confidence reached for beta {beta} and radius {radius}. Converged.')
+                    break
+            if (iteration >= stagnation_window + 1 and not stagnation_threshold == np.inf
+                    and all(np.isclose(
+                        [max(update_norms[-stagnation_window - 1:]), min(update_norms[-stagnation_window - 1:])],
+                        update_norm, atol=stagnation_threshold))):
```

Comment:
I changed it to a relative threshold now. The check here works differently than in `newton` though, because we don't want to check that the error is reduced, but rather that the update of the parameters stalls out.